File: Scheduling Algos/fcfs/nim/fcfs_sort.nim
import algorithm, strformat

type
  Process = object
    pid: Natural
    at: Natural
    bt: Natural
    ct: Natural
  
var prc = @[Process(pid: 0, at: 5, bt: 3, ct: 0), Process(pid: 1, at: 2, bt: 1, ct: 0),
            Process(pid: 2, at: 0, bt: 1, ct: 0), Process(pid: 3, at: 4, bt: 2, ct: 0),
            Process(pid: 4, at: 0, bt: 2, ct: 0), Process(pid: 5, at: 2, bt: 3, ct: 0)]
var ta = 0.0
var w = 0.0

proc findCompletionTime(): void =
  for i in countup(0, len(prc) - 1):
    if i == 0 or prc[i].at > prc[i - 1].ct:
      prc[i].ct = prc[i].at + prc[i].bt
    else:
      prc[i].ct = prc[i - 1].ct + prc[i].bt

prc.sort(proc (x, y: Process): int =
  if x.at > y.at: 1
  else: 0
)
findCompletionTime()
prc.sort(proc (x, y: Process): int =
  if x.pid > y.pid: 1
  else: 0
)

echo "PID\tAT\tBT\tCT\tTAT\tWT"
for p in prc:
  let pTA = (float)(p.ct - p.at)
  let pW = pTA - (float)(p.bt)
  echo &"{p.pid}\t{p.at}\t{p.bt}\t{p.ct}\t{(int)pTA}\t{(int)pW}"
  ta += pTA / (float)(len(prc))
  w += pW / (float)(len(prc))

echo &"Avg. TAT: {ta:.3f}"
echo &"Avg. WT: {w:.3f}"

File: Scheduling Algos/fcfs/python/fcfs-sort.py
def findCompletionTime():
    for i in range(len(proc)):
        if proc[i]['at'] > proc[i - 1]['ct']:
            proc[i]['ct'] = proc[i]['at'] + proc[i]['bt']
        else:
            proc[i]['ct'] = proc[i - 1]['ct'] + proc[i]['bt']

def findAvgTime():
    print('Avg TAT: ', end = '')
    print('%.3f'%(sum((x['ct'] - x['at']) for x in proc) / len(proc)))
    print('Avg WT: ', end = '')
    print('%.3f'%(sum((x['ct'] - (x['at'] + x['bt'])) for x in proc) / len(proc)))

if __name__ == '__main__':
    proc = [
            {'pid': 0, 'at': 2, 'bt': 5, 'ct': 0},
            {'pid': 1, 'at': 0, 'bt': 3, 'ct': 0},
            {'pid': 2, 'at': 1, 'bt': 6, 'ct': 0},
            {'pid': 3, 'at': 2, 'bt': 2, 'ct': 0},
            {'pid': 4, 'at': 7, 'bt': 1, 'ct': 0},
            {'pid': 5, 'at': 5, 'bt': 2, 'ct': 0}
            ]
    proc.sort(key = lambda x: x['at'])
    findCompletionTime()
    proc.sort(key = lambda x: x['pid'])

    print('PID\tAT\tBT\tCT\tTAT\tWT')
    [print(f"{p['pid']}",
        f"{p['at']}",
        f"{p['bt']}",
        f"{p['ct']}",
        f"{p['ct'] - p['at']}",
        f"{p['ct'] - (p['at'] + p['bt'])}",
        sep = '\t') for p in proc]
    findAvgTime()
